public/app/plugins/panel/geomap/layers/data/heatMap.ts

```diff
@@ -0,0 +1,202 @@
+        });
+        vectorLayer.setSource(vectorSource);
+
+        // Set gradient of heatmap
+        const colorMode = getFieldColorModeForField(field);
+        if (colorMode.isContinuous && colorMode.getColors) {
+          // getColors return an array of color string from the color scheme chosen
+          const colors = colorMode.getColors(theme);
+          vectorLayer.setGradient(colors);
+        } else {
+          // Set the gradient back to default if threshold or single color is chosen
+          vectorLayer.setGradient(['#00f', '#0ff', '#0f0', '#ff0', '#f00']);
+        }
+      },
+    };
+  },
+  // Heatmap overlay options
+  registerOptionsUI: (builder) => {
+    builder
+      .addSliderInput({
+        path: 'radius',
+        description: 'configures the size of clusters',
+        name: 'Radius',
+        defaultValue: defaultOptions.radius,
+        settings: {
+            min: 1,
+            max: 50,
+            step: 1,
+        },
+      })
+      .addSliderInput({
+        path: 'blur',
+        description: 'configures the amount of blur of clusters',
+        name: 'Blur',
+        defaultValue: defaultOptions.blur,
+        settings: {
+          min: 1,
+          max: 50,
+          step: 1,
+        },
+      })
+      .addSelect({
+        path: 'queryFormat.locationType',
+        name: 'Query Format',
+        defaultValue: defaultOptions.queryFormat.locationType,
+        settings: {
+          options: [
+            {
+              value: 'coordinates',
+              label: 'Coordinates',
+            },
+            {
+              value: 'geohash',
+              label: 'Geohash',
+            },
+          ],
+        },
+      })
+      .addTextInput({
+        path: 'fieldMapping.metricField',
+        name: 'Metric Field',
+        defaultValue: defaultOptions.fieldMapping.metricField,
+      })
+      .addTextInput({
+        path: 'fieldMapping.latitudeField',
+        name: 'Latitude Field',
+        defaultValue: defaultOptions.fieldMapping.latitudeField,
+        showIf: (config) =>
+          config.queryFormat.locationType === 'coordinates',
+      })
+      .addTextInput({
+        path: 'fieldMapping.longitudeField',
+        name: 'Longitude Field',
+        defaultValue: defaultOptions.fieldMapping.longitudeField,
+        showIf: (config) =>
+          config.queryFormat.locationType === 'coordinates',
+      })
+      .addTextInput({
+        path: 'fieldMapping.geohashField',
+        name: 'Geohash Field',
+        defaultValue: defaultOptions.fieldMapping.geohashField,
+        showIf: (config) =>
+          config.queryFormat.locationType === 'geohash',
+      });
+  },
+
+  // fill in the default values
+  defaultOptions,
+};
+
+/**
+ * Function that normalize the data values to a range between 0 and 1
+ * Returns the weights for each value input
+ */
+function normalize(calcs: FieldCalcs, value: number) {
+  return (value - calcs.min) / calcs.range;
```

Comment:
Can this range value be 0? Does that destroy anything?

Reply:
good point, it would not visualize the minimum data. 
I could check for the min value and return a small weight (ex. 0.01) but it seems like any weights below 0.1 is actually barely visible and would be better to scale it in a range of [0.1, 1].